### PR TITLE
Add app settings table

### PR DESCRIPTION
**Pull request by Flanker32**
Add app settings table for creation/local run ui component

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        return String.class;
+    }
+
+    @Override
+    public boolean isCellEditable(int i, int i1) {
+        return true;
+    }
+
+    @Override
+    public Object getValueAt(int i, int i1) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java]: row, column

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class AppSettingModel implements TableModel {
+
+    private static final String[] TITLE = {"Key", "Value"};
+
+    private List<Pair<String, String>> appSettings = new ArrayList<>();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java]: Use a Map<String, String> using LinkedHashMap?

Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java]: Map didn't provide interface to get value by index, so change to use List instead

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        JsonObject jsonObject = getLocalSettingsJsonObject(target);
+        if (jsonObject == null) {
+            jsonObject = gson.fromJson(DEFAULT_LOCAL_SETTINGS_JSON, JsonObject.class);
+        }
+        final JsonObject valueObject = new JsonObject();
+        appSettings.entrySet().forEach(entry -> valueObject.addProperty(entry.getKey(), entry.getValue()));
+        jsonObject.add(LOCAL_SETTINGS_VALUES, valueObject);
+        try (Writer writer = new FileWriter(target)) {
+            gson.toJson(jsonObject, writer);
+        } catch (IOException e) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: remover  catch (IOException e)

Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: Fixed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java (hunk 1 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java]: Renamed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+    @Override
+    public Object getValueAt(int row, int col) {
+        final Pair<String, String> target = appSettings.get(row);
+        return col == 0 ? target.getKey() : target.getValue();
+    }
+
+    @Override
+    public void setValueAt(Object value, int row, int col) {
+        while (row >= appSettings.size()) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java]: throw IAE error if value is not null and is not a string

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        }
+        final JsonObject valueObject = new JsonObject();
+        appSettings.entrySet().forEach(entry -> valueObject.addProperty(entry.getKey(), entry.getValue()));
+        jsonObject.add(LOCAL_SETTINGS_VALUES, valueObject);
+        try (Writer writer = new FileWriter(target)) {
+            gson.toJson(jsonObject, writer);
+        }
+    }
+
+    private static JsonObject getLocalSettingsJsonObject(File target) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: readJsonFile

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java (hunk 4 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingModel.java]: Added

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            return new HashMap<>();
+        }
+    }
+
+    public static void exportLocalSettingsJsonFile(File target, Map<String, String> appSettings) throws IOException {
+        if (target == null) {
+            return;
+        }
+        if (!target.exists()) {
+            target.createNewFile();
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: shall we check parent folder existence?

Comment (andxu) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: Path `xxx` does not exist, please check the path and try again.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+        final JsonObject valueObject = new JsonObject();
+        appSettings.entrySet().forEach(entry -> valueObject.addProperty(entry.getKey(), entry.getValue()));
+        jsonObject.add(LOCAL_SETTINGS_VALUES, valueObject);
+        try (Writer writer = new FileWriter(target)) {
+            gson.toJson(jsonObject, writer);
+        }
+    }
+
+    private static JsonObject readLocalSettingsJsonFile(File target) {
+        try (FileInputStream fis = new FileInputStream(target);
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: add a todo mark: we need to use `azure-maven-plugin-common/src/main/java/com/microsoft/azure/maven/common/utils/JsonUtils.java` when it is moved to azure tools common

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+public class AppSettingsTableUtils {
+
+    private static final String DEFAULT_LOCAL_SETTINGS_JSON = "{\"IsEncrypted\":false,\"Values\":{\"AzureWebJobsStorage\":\"\",\"FUNCTIONS_WORKER_RUNTIME\":\"java\"}}";
+    private static final String LOCAL_SETTINGS_VALUES = "Values";
+    private static final String EXPORT_APP_SETTINGS = "Export app settings to local.settings.json";
+    private static final String EXPORT_APP_SETTINGS_FAILED = "Fail to save app settings to local.settings.json";
+    private static final String LOCAL_SETTINGS_JSON = "local.settings.json";
+    private static final String EXPORT_LOCAL_SETTINGS_TITLE = "Export to local settings";
+
+    public static JPanel decorateAppSettingTable(AppSettingsTable appSettingsTable) {
```
Comment (andxu) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: createAppSettingTablePanel

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java (hunk 7 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: Done

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java (hunk 8 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: Renamed

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java (hunk 6 above)
Comment (Flanker32) [reply PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/intellij/runner/functions/component/table/AppSettingsTableUtils.java]: Added the check and will create parent folder if not exists

**Pull request merged**